src/civetweb.c

```diff
@@ -6031,68 +6084,117 @@ static int set_ports_option(struct mg_context *ctx)
-    } else {
-        mg_strlcpy(date, "01/Jan/1970:00:00:00 +0000", sizeof(date));
-        date[sizeof(date) - 1] = '\0';
+
+    if (conn->ctx->callbacks.log_access) {
+        conn->ctx->callbacks.log_access(conn, conn->ctx->config[ACCESS_LOG_FORMAT]);
     }
 
-    ri = &conn->request_info;
+    if (!fp) {
+        return;
+    }
 
-    sockaddr_to_string(src_addr, sizeof(src_addr), &conn->client.rsa);
-    referer = header_val(conn, "Referer");
-    user_agent = header_val(conn, "User-Agent");
+    mg_strlcpy(log_format, conn->ctx->config[ACCESS_LOG_FORMAT], sizeof(log_format));
 
-    snprintf(buf, sizeof(buf), "%s - %s [%s] \"%s %s HTTP/%s\" %d %" INT64_FMT " %s %s",
-            src_addr, ri->remote_user == NULL ? "-" : ri->remote_user, date,
-            ri->request_method ? ri->request_method : "-",
-            ri->uri ? ri->uri : "-", ri->http_version,
-            conn->status_code, conn->num_bytes_sent,
-	    referer, user_agent);
 
-    if (conn->ctx->callbacks.log_access) {
-        conn->ctx->callbacks.log_access(conn, buf);
+    ptm = gmtime_r(&conn->birth_time, ptm);
+    if (ptm != NULL) {
+        strftime(date, sizeof(date), "%Y-%m-%d", ptm);
+        strftime(time, sizeof(time), "%H-%M-%S", ptm);
+    } else {
+        mg_strlcpy(time, "00-00-00", sizeof(time));
+        mg_strlcpy(date, "1970-01-01", sizeof(date));
     }
+    sockaddr_to_string(src_addr, sizeof(src_addr), &conn->client.rsa);
+    snprintf(status_code_str, sizeof(status_code_str), "%d", conn->status_code);
+    snprintf(num_bytes_str, 50, "%"
+    INT64_FMT, conn->num_bytes_sent);
+    const char *field_values[] = {
+            "date", date,
+            "time", time,
+            "c-ip", src_addr,
+            "cs-username", ri->remote_user == NULL ? "-" : ri->remote_user,
+            "cs-method", ri->request_method ? ri->request_method : "-",
+            "cs-uri-query", ri->query_string ? ri->query_string : "-",
+            "cs-uri", ri->uri ? ri->uri : "-",
+            "sc-status", status_code_str,
+            "sc-bytes", num_bytes_str,
+            NULL //marker for end of array
+    };
 
-    if (fp) {
+    for (; (token = strtok_r(log_format_ptr, " \n\f\t\r\v", &saveptr)) && ok; log_format_ptr = NULL) {
+        const char *prefix = log_format_ptr == NULL ? " " : "";
+        const int prefix_len = strlen(prefix);
+        const int token_len = strlen(token);
+        const char *to_write = NULL;
+        // process request headers
+        if (strncmp("cs(", token, 3) == 0 && token[token_len - 1] == ')') {
+            char header_name[token_len];
+            strncpy(header_name, token + 3, token_len - 4);
+            header_name[token_len - 4] = '\0';
+            char *header_val = mg_get_header(conn, header_name);
+            if (header_val) {
+                char esc_hdr_val[MAX_LOG_FORMAT_LEN];
+                // escape quotes
+                ok &= mg_str_replace(esc_hdr_val, ARRAY_SIZE(esc_hdr_val), header_val, "\"", "\"\"");
```

Comment:
Escape quotes would mean convert this: ...x"y... to this ...x\"y... ?
Whereas _str_replace() seems to be doing this: ...x""y... ?